File: CandyCrush.js/candy-003.js
// 編號：CANDY-003
// 程式語言：JavaScript
// 題目：完成函數的內容，把陣列裡的 0 都移到最後面

// let list = [false, 1, 0, -1, 2, 0, 1, 3, "a"];

// function moveZerosToEnd(arr) {
//   let nonZero = [];
//   let zero = [];
//   for (i = 0; i < arr.length; i++) {
//     if (arr[i] !== 0) {
//       nonZero.push(arr[i]);
//     } else {
//       zero.push(arr[i]);
//     }
//   }
//   return nonZero.concat(zero);
// }
// let result = moveZerosToEnd(list);
// console.log(result); // 印出 [false, 1, -1, 2, 1, 3, "a", 0, 0]

// solving steps:
// 先宣告兩種新的空陣列，以 for 迴圈判斷不是 0 或者是 0，分別放入一個放"不是" 0 的，一個放"是" 0 的陣列中，為了將兩個陣列合在一起，我使用 concat() 方法做合併。

// 編號：CANDY-003 -- revision
// 程式語言：JavaScript
// 題目：完成函數的內容，把陣列裡的 0 都移到最後面

let list = [false, 1, 0, -1, 2, 0, 1, 3, "a"];

function moveZerosToEnd(arr) {
  let nonZero = arr.filter((element) => element !== 0);
  let zero = arr.filter((element) => element === 0);
  return nonZero.concat(zero);
}

let result = moveZerosToEnd(list);
console.log(result); // 印出 [false, 1, -1, 2, 1, 3, "a", 0, 0]

// solving steps:
// 以 .filter() 方法去過濾不是 0 或者是 0 得陣列，為了將兩個陣列合在一起，我使用 concat() 方法做合併。

File: CandyCrush.rb/candy-003.rb
# 編號：CANDY-003
# 程式語言：Ruby
# 題目：完成函數的內容，把陣列裡的 0 都移到最後面

list = [false, 1, 0, -1, 2, 0, 1, 3, "a"];

def moveZerosToEnd(arr)
  non_zero_list = arr.select {|element| element != 0}
  zero_list = arr.select {|element| element == 0}
  non_zero_list + zero_list
end


p moveZerosToEnd(list); # 印出 [false, 1, -1, 2, 1, 3, "a", 0, 0]

# solving steps:
# 設定兩個陣列一個放非零的、一個放是零的，最後把兩個陣列相加。